src/test/java/io/neonbee/test/listeners/StaleVertxChecker.java

```diff
@@ -0,0 +1,159 @@
+
+    private static final Method CONTEXT_METHOD;
+
+    static {
+        Method contextMethod = null;
+        try {
+            (contextMethod = VertxThread.class.getDeclaredMethod("context")).setAccessible(true);
+        } catch (NoSuchMethodException | SecurityException e) {
+            LOGGER.warn("Cannot set context method of VertxThread accessible, checking for stale threads is limited");
+        } finally {
+            CONTEXT_METHOD = contextMethod;
+        }
+    }
+
+    @Override
+    public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+        super.executionFinished(testIdentifier, testExecutionResult);
+        if (CONTEXT_METHOD != null) {
+            checkForStaleVertxInstances(testIdentifier);
+        }
+    }
+
+    @SuppressWarnings({ "PMD.EmptyIfStmt", "checkstyle:MultipleVariableDeclarations" })
+    private void checkForStaleVertxInstances(TestIdentifier testIdentifier) {
+        LOGGER.info("Checking for stale Vert.x instances");
+
+        // first try to determine all Vert.x instances that are currently available (and running?)
+        Set<Vertx> vertxInstances = findStaleThreads(VERTX_THREAD_NAME_PREFIX).filter(VertxThread.class::isInstance)
+                .map(VertxThread.class::cast).map(thread -> {
+                    try {
+                        Context context = (Context) CONTEXT_METHOD.invoke(thread);
+                        if (context == null) {
+                            LOGGER.debug("Vert.x thread {} is current not associated to any context", thread);
+                            return null;
+                        }
+
+                        Vertx vertx = context.owner();
+                        if (vertx == null) {
+                            LOGGER.debug("Vert.x thread {} has a context {} with no owner, is this a bug?!", context,
+                                    thread);
+                            return null;
+                        }
+
+                        return vertx;
+                    } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
+                        LOGGER.debug("Failed to determine Vert.x context for thread {}", thread);
+                        return null;
+                    }
+                }).collect(Collectors.toSet());
+
+        if (vertxInstances.isEmpty()) {
+            // perfect! we don't need to bother ourself!
+            LOGGER.info("No stale Vert.x instances found");
+            return;
+        } else if (vertxInstances.contains(null)) {
+            LOGGER.warn("Could not determine Vert.x instance for all Vert.x threads");
+        }
+
+        // try to find instances associated to this test or any instances that we cannot associate to any test
+        Set<Vertx> unassociatedVertxInstances = new HashSet<>(), associatedVertxInstances = new HashSet<>(), // NOPMD
+                notAssociatedVertxInstances = new HashSet<>();
```

Comment:
No, this does not make sense to me. Then we should adapt the PMD ruling.

If I want to define multiple sets with the same generic type, which MUST be of the same generic type, it doesn't make sense to declare it multiple times, which is the bigger code smell to me, because it assumes that the sets can be of different generic type, which is not true:

Set<Vertx> unassociatedVertxInstances = new HashSet<>();

Set<Vertx> associatedVertxInstances = new HashSet<>();

Set<Vertx> notAssociatedVertxInstances = new HashSet<>();

Thus it is in one declaration, because it makes it clear: "These sets MUST be compatible to each other".

Please let me know if the PMD rule should be adapted, or if you are fine with the NOPMD here.